src/mongoc/mongoc-server-description.c

```diff
@@ -0,0 +1,483 @@
+         is_shard = bson_iter_bool (&iter);
+      } else if (strcmp ("setName", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_UTF8 (&iter)) goto ERROR;
+
+         sd->set_name = bson_iter_utf8 (&iter, NULL);
+      } else if (strcmp ("secondary", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_secondary = bson_iter_bool (&iter);
+      } else if (strcmp ("hosts", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_ARRAY (&iter)) goto ERROR;
+         bson_iter_array (&iter, &len, &bytes);
+         bson_init_static (&sd->hosts, bytes, len);
+      } else if (strcmp ("passives", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_ARRAY (&iter)) goto ERROR;
+         bson_iter_array (&iter, &len, &bytes);
+         bson_init_static (&sd->passives, bytes, len);
+      } else if (strcmp ("arbiters", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_ARRAY (&iter)) goto ERROR;
+         bson_iter_array (&iter, &len, &bytes);
+         bson_init_static (&sd->arbiters, bytes, len);
+      } else if (strcmp ("primary", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_UTF8 (&iter)) goto ERROR;
+         sd->current_primary = bson_iter_utf8 (&iter, NULL);
+      } else if (strcmp ("arbiterOnly", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_arbiter = bson_iter_bool (&iter);
+         /*
+      } else if (strcmp ("passive", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_passive = bson_iter_bool (&iter);
+      } else if (strcmp ("hidden", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_hidden = bson_iter_bool (&iter);
+         */
+      } else if (strcmp ("isreplicaset", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_replicaset = bson_iter_bool (&iter);
+      } else if (strcmp ("tags", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_DOCUMENT (&iter)) goto ERROR;
+         bson_iter_document (&iter, &len, &bytes);
+         bson_init_static (&sd->tags, bytes, len);
+      } else if (strcmp ("me", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_UTF8 (&iter)) goto ERROR;
+         sd->connection_address = bson_iter_utf8 (&iter, NULL);
+      }
+   }
+
+   if (is_shard) {
+      sd->type = MONGOC_SERVER_MONGOS;
+   } else if (sd->set_name) {
+      if (is_master) {
+         sd->type = MONGOC_SERVER_RS_PRIMARY;
+      } else if (is_secondary) {
+         sd->type = MONGOC_SERVER_RS_SECONDARY;
+      } else if (is_arbiter) {
+         sd->type = MONGOC_SERVER_RS_ARBITER;
+      } else {
+         sd->type = MONGOC_SERVER_RS_OTHER;
+      }
+   } else if (is_replicaset) {
+      sd->type = MONGOC_SERVER_RS_GHOST;
+   } else if (is_master) {
```

Comment:
according to the SDAM spec, a server should have a type of STANDALONE if "No "msg: isdbgrid", no setName, and no "isreplicaset: true". - it isn't actually required to have `isMaster=true`.